### ISSUE TITLE
Explore single RNN-like module for *all* neuron dynamics

### ISSUE THREAD

**Issue by Jegp** [norse/norse#166]
PyTorch has a single [RNN base module](https://pytorch.org/docs/stable/_modules/torch/nn/modules/rnn.html#RNNBase) that applies all the recurrent activation dynamics in time. We should first study how the module works (why are they using strings as activations and not callables?!) and how they shove gradients around (#163) and then build a similar construction for Norse as discussed in #121.

Here's an ad hoc checklist:
- [x] Build a Norse version of the RNN Base
- [x] Add thorough tests for functionality
- [x] Add thorough tests for gradients in *multiple* connected layers

**Comment by Jegp** [comment]
This is also interesting for inspiration: https://keras.io/api/layers/recurrent_layers/time_distributed/

**Comment by Huizerd** [comment]
I have an idea for this, which would also tackle #169 and #143 (and #29).

Not sure it's fully working, but most of it should be. I think it decreases the 'thickness' of the layer on top of PyTorch that is Norse, and gives the user a bit more flexibility while still being useful. I also implemented heterogeneous/learnable time constants. Please have a look!

```python
from typing import NamedTuple, Optional, Tuple

import torch
import torch.nn as nn


class SpikeFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, v):
        ctx.save_for_backward(v)
        return v.gt(0.0).float()

    @staticmethod
    def backward(ctx, grad_output):
        (v,) = ctx.saved_tensors
        grad_input = grad_output.clone()
        spike_pseudo_grad = 0.3 * torch.nn.functional.relu(1 - v.abs())
        return grad_input * spike_pseudo_grad, None


def threshold(v: torch.Tensor, *args) -> torch.Tensor:
    return SpikeFunction.apply(v)


class OptionalLearnable(NamedTuple):
    tensor: torch.Tensor
    learnable: bool


class LIFParameters(nn.Module):
    def __init__(
        self,
        tau_syn_inv: OptionalLearnable = OptionalLearnable(tensor=torch.tensor(1.0 / 5e-3), learnable=False),
        tau_mem_inv: OptionalLearnable = OptionalLearnable(tensor=torch.tensor(1.0 / 1e-2), learnable=False),
        v_leak: torch.Tensor = torch.tensor(0.0),
        v_th: torch.Tensor = torch.tensor(1.0),
        v_reset: torch.Tensor = torch.tensor(0.0),
        method: str = "super",
        alpha: torch.Tensor = torch.tensor(100.0),
    ):
        super().__init__()
        if tau_syn_inv.learnable:
            self.tau_syn_inv = nn.Parameter(tau_syn_inv.tensor)
        else:
            self.register_buffer("tau_syn_inv", tau_syn_inv.tensor)
        if tau_mem_inv.learnable:
            self.tau_mem_inv = nn.Parameter(tau_mem_inv.tensor)
        else:
            self.register_buffer("tau_mem_inv", tau_mem_inv.tensor)
        self.register_buffer("v_leak", v_leak)
        self.register_buffer("v_th", v_th)
        self.register_buffer("v_reset", v_reset)
        self.method = method
        self.register_buffer("alpha", alpha)


class LIFState(NamedTuple):
    z: torch.Tensor
    v: torch.Tensor
    i: torch.Tensor


class LIFCell(nn.Module):
    def __init__(
        self,
        p: LIFParameters = LIFParameters(),
        dt: float = 0.001,
    ):
        super().__init__()
        self.p = p
        self.dt = dt

    def forward(
        self,
        input_tensor: torch.Tensor,
        state: Optional[torch.Tensor] = None,
    ) -> Tuple[torch.Tensor, LIFState]:
        if state is None:
            state = LIFState(
                z=torch.zeros_like(input_tensor),
                v=torch.ones_like(input_tensor) * self.p.v_leak,  # why was here a .detach?
                i=torch.zeros_like(input_tensor),
            )
        return lif_step(input_tensor, state, self.p, self.dt)


def lif_step(
    input_tensor: torch.Tensor,
    state: torch.Tensor,
    p: LIFParameters,
    dt: float,
) -> Tuple[torch.Tensor, LIFState]:
    # One could argue current update before voltage update?
    dv = dt * p.tau_mem_inv * ((p.v_leak - state.v) + state.i)
    v_decayed = state.v + dv

    di = -dt * p.tau_syn_inv * state.i
    i_decayed = state.i + di

    z_new = threshold(v_decayed - p.v_th, p.method, p.alpha)
    v_new = (1 - z_new) * v_decayed + z_new * p.v_reset
    i_new = i_decayed + input_tensor  # only difference here

    return z_new, LIFState(z_new, v_new, i_new)


class DummyEncoder(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return x


class NonSpikingDecoder(nn.Module):
    def __init__(self, in_size: int, out_size: int, tau_mem: OptionalLearnable = OptionalLearnable(tensor=torch.tensor(20.0), learnable=False)):
        super().__init__()

        # For this decoder I just put the synapse here, but it can of course also be put in RecurrentSNN for consistency with the layers
        self.synapse = nn.Linear(in_size, out_size)
        if tau_mem.learnable:
            self.tau_mem = nn.Parameter(tau_mem.tensor)
        else:
            self.register_buffer("tau_mem", tau_mem)
    
    def forward(self, x: torch.Tensor, v: Optional[torch.Tensor] = None) -> torch.Tensor:
        if v is None:
            v = torch.zeros_like(self.synapse(x))
        # I just put simplified dynamics here..
        v = v * torch.exp(1.0 / self.tau_mem) + self.synapse(x)
        return v
        

class RecurrentSNN(nn.Module):
    def __init__(self):
        super().__init__()

        # Real -> spikes and vice-versa
        self.encoder = DummyEncoder()
        self.decoder = NonSpikingDecoder(128, 10)

        # Synapses
        self.ff1 = nn.Linear(1, 256)
        self.rec1 = nn.Linear(256, 256)
        self.ff2 = nn.Linear(256, 128)
        self.rec2 = nn.Linear(128, 128)

        # Neurons
        p1 = LIFParameters(
            tau_syn_inv=OptionalLearnable(tensor=torch.normal(200, 25, size=[256]), learnable=True),
            tau_mem_inv=OptionalLearnable(tensor=torch.normal(100, 25, size=[256]), learnable=True),
        )
        p2 = LIFParameters(
            tau_syn_inv=OptionalLearnable(tensor=torch.normal(100, 10, size=[128]), learnable=True),
            tau_mem_inv=OptionalLearnable(tensor=torch.normal(500, 10, size=[128]), learnable=True),
        )
        self.lif1 = LIFCell(p1)
        self.lif2 = LIFCell(p2)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        # Assume x has shape (batch, seq, ...)
        _, seq, _ = x.shape
        # Encode all at once
        x = self.encoder(x)

        # Then loop
        state1, state2, v = None, None, None
        z1 = torch.zeros(256, device=x.device, dtype=x.dtype)
        z2 = torch.zeros(128, device=x.device, dtype=x.dtype)
        for i in seq:
            z1, state1 = self.lif1(self.ff1(x[:, i]) + self.rec1(z1), state1)
            z2, state2 = self.lif2(self.ff2(z1) + self.rec2(z2), state2)
            v = self.decoder(z2, v)
        
        return v
```

**Comment by Jegp** [comment]
Great input @Huizerd! I like that the RNN is grouped in a single module, the neat modularization, and the seamless PyTorch compatibility (as you wrote). I'm a particular fan of your Parameters class to toggle parameter learning. 
Here are a few comments/questions
- This is purely LIF, and it would probably make sense to parameterize the dynamics. Exactly how is unclear to me, though. [PyTorch chose a pretty weird way of doing this](https://pytorch.org/docs/stable/_modules/torch/nn/modules/rnn.html#RNNBase) by string comparison in a super class, but it might be because it plays well with the C++/CUDA backends.
- Would you embed the encoder/decoder in the module? What if you want to inject event-based data? Or put it in serial with other modules?
- You assume data is shaped `(batch, seq, ...)`. It seems reasonable to me, since this is how data loaders would output the data. But are there use cases where this is not the case?
- The `LIFState` is giving me some headaches because it inherits from the `NamedTuple` and technically isn't a `Tensor`. I'm having my eye out for the [NamedTensor](https://pytorch.org/docs/stable/named_tensor.html). It's still experimental but what do you think about migrating to that?

**Comment by Huizerd** [comment]
Actually, the RNN was just to show how users would create their own RNNs, while Norse would only supply the dynamics of the neurons + encoders/decoders.

I looked at NamedTensor, and it seems like it only supports naming dimensions of a single tensor, not grouping multiple tensors together with names?

**Comment by Jegp** [comment]
That makes sense. I think we should also pursue some form of singular RNNBase-like module to capture a lot of the time sequence thing. But I think this is a nice foundation :-)

Yeah, I'm slightly hoping to merge the state into something simpler than a NamedTuple (single tensor or perhaps just a tuple of tensors) for three reasons: 1) it's much easier to operate on (send to device, transform, broadcast, etc.), 2) it replicates the PyTorch RNN/LSTM/GRU state which is also tensors, which allows us to do checkpointing and other optimizations, 3) we can dramatically reduce code because we don't have to hardcode all the tuple objects. Admittedly, that's also the downside: it'll be much easier to insert wrong state and get weird errors. But that could be mitigated by adding proper error handling.

**Comment by Huizerd** [comment]
A single tensor for state could be nice. [Here](https://github.com/pytorch/pytorch/blob/master/benchmarks/fastrnns/custom_lstms.py#L108) they cut the tensor into 4 parts using `.chunk(4)`. I guess we could do 
```python
state = torch.zeros(3, size)  # to create tensor
z, v, i = state.chunk(3, dim=0)  # to split state and use parts
```

**Comment by Jegp** [comment]
Nice find! That is pretty easy indeed. If you think this not entirely off, I might want to start implementing this in the coming week. Any help would be greatly appreciated of course :-)

**Comment by Huizerd** [comment]
Sure! Just let me know if there's anything I can do

**Comment by Jegp** [comment]
You're also more than welcome to do an initial draft. Alternatively, we can start it out together on a zoom peer coding session or something. That'll be both fun and productive, I think. Otherwise I expect to have something for the weekend.

**Comment by Huizerd** [comment]
It would be nice to see your suggestions, maybe we can set up a zoom session after that (this week is kind of busy for me :))

**Comment by cpehle** [comment]
I would be happy to join as well. I didn't chime in so far because I'm also very busy this week (or for that matter always :)).
